lib/s3-cache-features.js

```diff
@@ -0,0 +1,328 @@
+'use strict';
+
+var __ = require('lodash');
+var async = require('async');
+var concat = require('concat-stream');
+var projector = require("nodetiles-core").projector;
+var Promise = require('bluebird');
+var xml2js = require('xml2js');
+
+var metrics = require('./metrics/cache');
+var Response = require('./models/Response');
+var settings = require('./settings');
+
+var knex = require('knex');
+
+// XXX TODO -- should this connection be created at the app level?
+var pg = knex({
+  client: 'pg',
+  connection: settings.psqlConnectionString
+});
+
+
+var S3_CUSTOM_CACHE_HEADER = 'x-amz-meta-count';
+
+function tile2long(x, z) {
+  return (x/Math.pow(2,z)*360-180);
+}
+
+function tile2lat(y, z) {
+  var n=Math.PI-2*Math.PI*y/Math.pow(2,z);
+  return (180 / Math.PI * Math.atan(0.5 * (Math.exp(n) - Math.exp(-n))));
+}
+
+function tileToBounds(tile) {
+  var sw = [tile2long(tile[1], tile[0]), tile2lat(tile[2] + 1, tile[0])];
+  var ne = [tile2long(tile[1] + 1, tile[0]), tile2lat(tile[2], tile[0])];
+  return [sw, ne];
+}
+
+module.exports = function setup(options) {
+  var s3client = options.s3client;
+
+  function makeEnvelope(tile) {
+    var bounds = tileToBounds(tile);
+    var west = bounds[0][0];
+    var south = bounds[0][1];
+    var east = bounds[1][0];
+    var north = bounds[1][1];
+
+    return knex.raw('geom && ST_MakeEnvelope(?, ?, ?, ?)', [west, south, east, north]);
+  }
+
+  function getCount(tile) {
+    var query = pg.count('object_id')
+      .from(settings.featuresTable)
+      .where(makeEnvelope(tile));
+
+    return query;
+  }
+
+  function hasNew(tile, timestamp) {
+    // Convert the timestamp to UNIX time (what postgres wants)
+    timestamp = new Date(timestamp).getTime() / 1000;
+
+    var query = pg.select('timestamp')
+      .from(settings.featuresTable)
+      .where(makeEnvelope(tile))
+      .andWhere(knex.raw("timestamp > to_timestamp(?)", [timestamp]))
+      .limit(1);
+
+    return query;
+  }
+
+  function getS3Header(name, done) {
+    s3client.head(name)
+    .on('response', function (s3res) {
+      done(null, s3res);
+    })
+    .on('error', done)
+    .end();
+  }
+
+  function getS3Object(name, done) {
+    s3client.get(name)
+    .on('response', function (s3res) {
+      done(null, s3res);
+    })
+    .on('error', done)
+    .end();
+  }
+
+  var getS3ObjectAsync = Promise.promisify(getS3Object);
+
+  var makeTileName = (function () {
+    var prefix = '/' + settings.name;
+    return function makeTileName(req) {
```

Comment:
This is a mechanism we've had to tweak before, to solve a bug or two related to caching and safe names. It would be good to factor it out of the two S3 cache modules, in case we need to tweak it later.